### ISSUE TITLE
arch: posix: _POSIX_C_SOURCE declaration conflicts with the rest of the system

### ISSUE THREAD

**Issue by cfriedt** [zephyrproject-rtos/zephyr#66921]
**Describe the bug**
Picolibc and arch/posix both globally define a number of symbols reserved by IEEE-1003.1-2017 (and many earlier versions of the standard) such as `_POSIX_C_SOURCE`, `_XOPEN_SOURCE` for all Zephyr builds on that architecture, even Zephyr builds which do not explicitly want or use `CONFIG_POSIX_API=y` or any POSIX [subprofiling options](https://docs.zephyrproject.org/latest/services/portability/posix/option_groups/index.html).

This is a bit of an issue, because 
a) it means virtually every platform is in violation of [Rules A.4 and A.5 of the Coding Guidelines](https://github.com/zephyrproject-rtos/zephyr/issues/66921) since
  * ~2018 for arch/posix
  * ~2020 (based on commit date) for lib/libc/picolibc
b) since the compile definitions are hard-coded and global, they prevent applications from defining (correct, standard) application conformance levels without risking a compiler error

<img width="823" alt="Screenshot 2023-12-27 at 8 49 04 PM" src="https://github.com/zephyrproject-rtos/zephyr/assets/983494/c85e2dec-1bd3-406a-adfa-2fe3078052dd">

<img width="772" alt="Screenshot 2023-12-27 at 9 22 21 PM" src="https://github.com/zephyrproject-rtos/zephyr/assets/983494/0dfa6ed9-26da-4406-b644-7894e987bab2">

It seems to only be an issue since `native_sim` has facilitated running applications that enable features defined by Zephyr's POSIX API for approximately 6 months.

Please also mention any information which could help others to understand
the problem you're facing:
 - What target platform are you using? all
 - What have you tried to diagnose or workaround this issue? PRs, etc
 - Is this a regression? No

**To Reproduce**

Attempt to build any application conforming to the IEEE 1003.1-1-2017 standard with either up-to-date `_POSIX_C_SOURCE` (200809L) or up-to-date `_XOPEN_SOURCE` (700). (note: 
1.
```shell
echo -e "#define _POSIX_C_SOURCE 200809L\n#define _XOPEN_SOURCE 700" \
  | cat - samples/hello_world/src/main.c > /tmp/foo.c
mv /tmp/foo.c samples/hello_world/src/main.c
```
2. `west build -b native_sim_64 samples/hello_world`
2. Observe errors.

**Expected behavior**
Ideally the internal compiler definitions used for `ARCH_POSIX` and Picolibc would not spill out and affect the rest of the build. 

POSIX application conformance macros should only be local compile definitions, and not global compile definitions.

**Impact**
Originally this had not been noticed until after `native_sim` was added, since it was not supported to enable `CONFIG_POSIX_API` and `CONFIG_ARCH_POSIX` together for many years. They were mutually exclusive due to architectural issues.

However, now that POSIX_API is ramping up significantly and nearing feature-completion, we are seeing more frequent conflicts. Originally, this was only an issue on `native_sim` and `native_sim_64`, but now seems to be affecting all platforms, since it is a bit of a showstopper for the [POSIX Roadmap for LTSv3](https://github.com/zephyrproject-rtos/zephyr/issues/51211).

**Logs and console output**
```shell
src/main.c.obj -MF CMakeFiles/app.dir/src/main.c.obj.d -o CMakeFiles/app.dir/src/main.c.obj -c /tmp/zephyrproject/zephyr/samples/hello_world/src/main.c
/tmp/zephyrproject/zephyr/samples/hello_world/src/main.c:1: error: "_POSIX_C_SOURCE" redefined [-Werror]
    1 | #define _POSIX_C_SOURCE 200809L
      | 
<command-line>: note: this is the location of the previous definition
/tmp/zephyrproject/zephyr/samples/hello_world/src/main.c:2: error: "_XOPEN_SOURCE" redefined [-Werror]
    2 | #define _XOPEN_SOURCE 700
```

**Environment (please complete the following information):**
 - OS: Linux
 - Toolchain: Zephyr SDK v0.16.4
 - Commit: 87d056bd31b14b680e6f011134c7ce433c4462f6

**Additional context**
Related PRs (IIRC, Newlib and Picolibc might also have issues here)
https://github.com/zephyrproject-rtos/zephyr/pull/67040

**Comment by cfriedt** [comment]
It seems that this issue is also not avoided by using `zephyr_library_compile_definitions()`. There was a hypothesis that this would limit compile definitions to the library being built, but as of now, it's quite easy to demonstrate broken on `main`.

The same issue exists if `zephyr_compile_definitions()` are used in both locations.
```diff
diff --git a/arch/posix/CMakeLists.txt b/arch/posix/CMakeLists.txt
index dc83cecc61..8b7bd8ffe6 100644
--- a/arch/posix/CMakeLists.txt
+++ b/arch/posix/CMakeLists.txt
@@ -133,7 +133,7 @@ if (CONFIG_GPROF)
   target_link_options(native_simulator INTERFACE "-pg")
 endif()
 
-zephyr_compile_definitions(_POSIX_C_SOURCE=200809 _XOPEN_SOURCE=600 _XOPEN_SOURCE_EXTENDED)
+zephyr_library_compile_definitions(_POSIX_C_SOURCE=200809 _XOPEN_SOURCE=600 _XOPEN_SOURCE_EXTENDED)
 
 if (CONFIG_NATIVE_APPLICATION)
   zephyr_ld_options(
diff --git a/lib/posix/CMakeLists.txt b/lib/posix/CMakeLists.txt
index c6b3a88aea..d1c2bc391f 100644
--- a/lib/posix/CMakeLists.txt
+++ b/lib/posix/CMakeLists.txt
@@ -35,6 +35,7 @@ if(CONFIG_POSIX_API OR CONFIG_PTHREAD_IPC OR CONFIG_POSIX_CLOCK OR
 endif()
 
 zephyr_library()
+zephyr_library_compile_definitions(_POSIX_C_SOURCE=200809L _XOPEN_SOURCE=700)
```

```shell
west build -p auto -b native_sim_64 -t run tests/posix/common/
```
<img width="750" alt="Screenshot 2023-12-27 at 9 17 31 PM" src="https://github.com/zephyrproject-rtos/zephyr/assets/983494/e5d69786-4252-4d15-820c-e56c7b6c2a56">

**Comment by aescolar** [comment]
> It seems that this issue is also not avoided by using `zephyr_library_compile_definitions()`. There was a hypothesis that this would limit compile definitions to the library being built, but as of now, it's quite easy to demonstrate broken on `main`.

`zephyr_library_compile_definitions()` works fine. You need to
a) Also change the line setting those macros for PICOLIBC
b) For the posix arch cmake, to set it when you are in the scope of the library, not the whole Zephyr scope/not when  `ZEPHYR_CURRENT_LIBRARY` is the whole Zephyr.
 
```diff
diff --git a/arch/posix/CMakeLists.txt b/arch/posix/CMakeLists.txt
index dc83cecc61..7001e54e75 100644
--- a/arch/posix/CMakeLists.txt
+++ b/arch/posix/CMakeLists.txt
@@ -133,8 +133,6 @@ if (CONFIG_GPROF)
   target_link_options(native_simulator INTERFACE "-pg")
 endif()
 
-zephyr_compile_definitions(_POSIX_C_SOURCE=200809 _XOPEN_SOURCE=600 _XOPEN_SOURCE_EXTENDED)
-
 if (CONFIG_NATIVE_APPLICATION)
   zephyr_ld_options(
     -ldl
diff --git a/arch/posix/core/CMakeLists.txt b/arch/posix/core/CMakeLists.txt
index 3cf83c9fdd..06640d0675 100644
--- a/arch/posix/core/CMakeLists.txt
+++ b/arch/posix/core/CMakeLists.txt
@@ -1,6 +1,7 @@
 # SPDX-License-Identifier: Apache-2.0
 
 zephyr_library()
+zephyr_library_compile_definitions(_POSIX_C_SOURCE=200809 _XOPEN_SOURCE=600 _XOPEN_SOURCE_EXTENDED)
 zephyr_library_compile_definitions(NO_POSIX_CHEATS)
 zephyr_library_sources(
        cpuhalt.c
diff --git a/lib/libc/picolibc/CMakeLists.txt b/lib/libc/picolibc/CMakeLists.txt
index 232e4ba8fc..5811a9cefa 100644
--- a/lib/libc/picolibc/CMakeLists.txt
+++ b/lib/libc/picolibc/CMakeLists.txt
@@ -12,7 +12,7 @@ if(NOT CONFIG_PICOLIBC_USE_MODULE)
   # Use picolibc provided with the toolchain
 
   zephyr_compile_options(--specs=picolibc.specs)
-  zephyr_compile_definitions(_POSIX_C_SOURCE=200809)
+  zephyr_library_compile_definitions(_POSIX_C_SOURCE=200809)
   zephyr_libc_link_libraries(--specs=picolibc.specs c -lgcc)
   if(CONFIG_PICOLIBC_IO_FLOAT)
     zephyr_compile_definitions(PICOLIBC_DOUBLE_PRINTF_SCANF)
diff --git a/lib/posix/CMakeLists.txt b/lib/posix/CMakeLists.txt
index c6b3a88aea..d1c2bc391f 100644
--- a/lib/posix/CMakeLists.txt
+++ b/lib/posix/CMakeLists.txt
@@ -35,6 +35,7 @@ if(CONFIG_POSIX_API OR CONFIG_PTHREAD_IPC OR CONFIG_POSIX_CLOCK OR
 endif()
 
 zephyr_library()
+zephyr_library_compile_definitions(_POSIX_C_SOURCE=200809L _XOPEN_SOURCE=700)
 add_subdirectory_ifdef(CONFIG_GETOPT getopt)
 zephyr_library_sources_ifdef(CONFIG_EVENTFD eventfd.c)
 zephyr_library_sources_ifdef(CONFIG_FNMATCH fnmatch.c)
```

**Comment by aescolar** [comment]
Removing the regression label as this has been the case since 2017. So it is not something that was working that broke due to some recent commit.

And clarifying that the issue is that if somebody wants to set _POSIX_C_SOURCE to 200809L they find themselves forced to set it to _POSIX_C_SOURCE=200809. So the assessment would be at an annoyance level, as the worst case is that somebody has some source setting this macro, and they need to either undef _POSIX_C_SOURCE first, or not set it in their code, or set it to the same value.

**Comment by cfriedt** [comment]
Just setting this to medium (it was high), because it's affecting every platform, causes violations of [rules A.4 and A.5](https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html) for just about every platform, and the release is coming up.

I would prefer to make this `priority: high` (and it was originally set to that 1 month ago).

Can we have a status update as to when this will be fixed?

**Comment by aescolar** [comment]
> Can we have a status update as to when this will be fixed?

@cfriedt currently you are blocking both PRs that are aiming at fixing this.

**Comment by cfriedt** [comment]
> > Can we have a status update as to when this will be fixed?
> 
> @cfriedt currently you are blocking both PRs that are aiming at fixing this. So if you unblock them it should be faster.

@aescolar - there are no linked PRs that seem to fix this issue. I have commented on https://github.com/zephyrproject-rtos/zephyr/pull/67040 and suggested to use local compile definitions instead of global compile definitions.

Is there another PR that should be linked to this issue?
